arch/common/multilevel_irq.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2018 Intel Corporation.
+ * Copyright (c) 2023 Meta.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/sw_isr_table.h>
+#include <zephyr/sys/util.h>
+
+/*
+ * Insert code if the node_id is an interrupt controller
+ */
+#define Z_IF_DT_IS_INTC(node_id, code)                                                             \
+	IF_ENABLED(DT_NODE_HAS_PROP(node_id, interrupt_controller), (code))
+
+/*
+ * Expands to node_id if its IRQN is equal to `irq`, nothing otherwise
+ * This only works for `irq` between 0 & 4095, see `IS_EQ`
+ */
+#define Z_IF_DT_INTC_IRQN_EQ(node_id, irq) IF_ENABLED(IS_EQ(DT_IRQ(node_id, irq), irq), (node_id))
+
+/*
+ * Expands to node_id if it's an interrupt controller & its IRQN is `irq`, or nothing otherwise
+ */
+#define Z_DT_INTC_GET_IRQN(node_id, irq)                                                           \
+	Z_IF_DT_IS_INTC(node_id, Z_IF_DT_INTC_IRQN_EQ(node_id, irq))
+
+/**
+ * Loop through child of "/soc" and get root interrupt controllers with `irq` as IRQN,
+ * this assumes only one device has the IRQN
+ * @param irq irq number
+ * @return node_id(s) that has the `irq` number, or empty if none of them has the `irq`
+ */
+#define INTC_DT_IRQN_GET(irq)                                                                      \
+	DT_FOREACH_CHILD_STATUS_OKAY_VARGS(DT_PATH(soc), Z_DT_INTC_GET_IRQN, irq)
+
+/* If can't find any matching interrupt controller, fills with `NULL` */
+#define INTC_DEVICE_INIT(node_id) .dev = DEVICE_DT_GET_OR_NULL(node_id),
+
+#define INIT_IRQ_PARENT_OFFSET(d, i, o) { \
+	INTC_DEVICE_INIT(d) \
+	.irq = i, \
+	.offset = o, \
+}
+
+#define IRQ_INDEX_TO_OFFSET(i, base) (base + i * CONFIG_MAX_IRQ_PER_AGGREGATOR)
+
+#define CAT_2ND_LVL_LIST(i, base) \
+	INIT_IRQ_PARENT_OFFSET(INTC_DT_IRQN_GET(CONFIG_2ND_LVL_INTR_0##i##_OFFSET), \
+			       CONFIG_2ND_LVL_INTR_0##i##_OFFSET, IRQ_INDEX_TO_OFFSET(i, base))
+const struct _irq_parent_entry _lvl2_irq_list[CONFIG_NUM_2ND_LEVEL_AGGREGATORS]
+	= { LISTIFY(CONFIG_NUM_2ND_LEVEL_AGGREGATORS, CAT_2ND_LVL_LIST, (,),
+		CONFIG_2ND_LVL_ISR_TBL_OFFSET) };
+
+#define CAT_3RD_LVL_LIST(i, base) \
+	INIT_IRQ_PARENT_OFFSET(INTC_DT_IRQN_GET(CONFIG_3RD_LVL_INTR_0##i##_OFFSET), \
+			       CONFIG_3RD_LVL_INTR_0##i##_OFFSET, IRQ_INDEX_TO_OFFSET(i, base))
+
+#ifdef CONFIG_3RD_LEVEL_INTERRUPTS
+
+const struct _irq_parent_entry _lvl3_irq_list[CONFIG_NUM_3RD_LEVEL_AGGREGATORS]
```

Comment:
Does this need to be conditionally defined?

E.g. can we remove the ifdef above, use `if (IS_ENABLED())` elsewhere, and let the linker discard unused code and data?

Reply:
yeah it fails to compile as the initialization used some undefined macros when 3rd level isn't enabled